File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    vector<pair<int,int>>dirc={{0,1},{1,0},{-1,0},{0,-1}};
    int orangesRotting(vector<vector<int>>& grid) {
        int m=grid.size();
        int n=grid[0].size();

        int maxi=0;
        queue<pair<int,pair<int,int>>>q;
            for(int i=0;i<m;i++){
                for(int j=0;j<n;j++){
                    if(grid[i][j]==2)q.push({0,{i,j}});
                }
            }

            while(!q.empty()){
                int time=q.front().first;
                int row=q.front().second.first;
                int col=q.front().second.second;
                q.pop();
                maxi=max(maxi,time);
                for(auto it:dirc){
                    int r=it.first+row;
                    int c=it.second+col;

                    if(r>=0 && r<m && c>=0 && c<n && grid[r][c]==1){
                        grid[r][c]=2;
                        q.push({time+1,{r,c}});
                    }
                }
            }

            for(auto it:grid){
                for(auto i:it){
                    if(i==1)return -1;
                }
            }

            return maxi;
    }
};

File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        //bfs
        int time =0;
        int [][]dirc={{1,0},{0,1},{-1,0},{0,-1}};
        int n=grid.length;
        int m=grid[0].length;
        int [][]vis=new int[n][m];
        

        Queue<int[]>q = new LinkedList<>();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                vis[i][j]=0;
                if(grid[i][j]==2){
                q.add(new int[]{i,j});
                vis[i][j]=1;}
            }
        }
        while(!q.isEmpty()){
            int len=q.size();

            for(int i=0;i<len;i++){
                int r=q.peek()[0];
                int c=q.peek()[1];

                q.remove();
                for(int []it:dirc){
                    int row=r+it[0];
                    int col=c+it[1];

                    if( row>=0 && row<n && col >=0 && col<m && vis[row][col]==0 && grid[row][col]==1){
                        vis[row][col]=1;
                        grid[row][col]=2;
                        q.add(new int[]{row,col});
                    }
                }
            }
             time++;
        }

         for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1 && vis[i][j]==0){
                   return -1;
            }
        }
         }
         if(time == 0) return 0; 
        return time-1;
}
}